doc/CGROUP_DELEGATION.md

```diff
@@ -0,0 +1,420 @@
+comprehensive up-to-date information about all this, particular in light of the
+poor implementations of the components interfacing with systemd of current
+container managers.
+
+Before you read on, please make sure you read the low-level [kernel
+documentation about
+cgroupsv2](https://www.kernel.org/doc/Documentation/cgroup-v2.txt). This
+documentation then adds in the higher-level view from systemd.
+
+This document augments the existing documentation we already have:
+
+* [The New Control Group Interfaces](https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/)
+* [Writing VM and Container Managers](https://www.freedesktop.org/wiki/Software/systemd/writing-vm-managers/)
+
+These wiki documents are not as up to date as they should be, currently, but
+the basic concepts still fully apply. You should read them too, if you do something
+with cgroups and systemd, in particular as they shine more light on the various
+D-Bus APIs provided. (That said, sooner or later we should probably fold that
+wiki documentation into this very document, too.)
+
+## Two Key Design Rules
+
+Much of the philosophy behind these concepts is based on a couple of basic
+design ideas of cgroupsv2 (which we however try to adapt as far as we can to
+cgroupsv1 too). Specifically two cgroupsv2 rules are the most relevant:
+
+1. The **no-processes-in-inner-nodes** rule: this means that it's not permitted
+to have processes directly attached to a cgroup that also has child cgroups and
+vice versa. A cgroup is either an inner node or a leaf node of the tree, and if
+it's an inner node it may not contain processes directly, and if it's a leaf
+node then it may not have child cgroups. (Note that there are some minor
+exceptions to this rule, though. i.e. the root cgroup is special and allows
+both processes and children — which is used in particular to maintain kernel
+threads.)
+
+2. The **single-writer** rule: this means that each cgroup only has a single
+writer, i.e. a single process managing it. It's OK if different cgroups have
+different processes managing them. However, only a single process should own a
+specific cgroup, and when it does that ownership is exclusive, and nothing else
+should manipulate it at the same time. This rule ensures that various pieces of
+software don't step on each other's toes constantly.
+
+These two rules have various effects. For example, one corollary of this is: if
+your container manager creates and manages cgroups in the system's root cgroup
+you violate rule #2, as the root cgroup is managed by systemd and hence off
+limits to everybody else.
+
+Note that rule #1 is generally enforced by the kernel if cgroupsv2 is used: as
+soon as you add a process to a cgroup it is ensured the rule is not
+violated. On cgroupsv1 this rule didn't exist, and hence isn't enforced, even
+though it's a good thing to follow it then too. Rule #2 is not enforced on
```

Comment:
not a blocker: `Rule #2 is not enforced` -> I wonder if that's something that can still be fixed in cgroupsv2... I was thinking of kernel adding a new file (`cgroup.manager`?) which would contain the pid of the single writer. If it's empty, then no one has made a claim on it and it's unclear who manages it. systemd (or other managers) could then implement policies on top of that by making claim to the cgroups they create, and either clearing that or switching pids when they delegate. Not sure if something like that was previously discussed. I think having a way to enforce this rule would go a long way in *preventing* container manager software from breaking this model. 

Also: `this is UNIX after all, root can do anything` -> not really *anything*, there are things like SELinux that can prevent root from doing stuff... Even without it, there are limits.

This discussion probably belongs elsewhere... So sorry for the tangential here. Your docs look good. They're a good description of the current state of the world.

Reply:
yeah @htejun originally wanted a scheme like that, but didn't implement that after all, as there was no nice obvious way how one could make that happen.

I discussed with @htejun now that it might make sense to add some limited, friendly log messaging if systemd notices that something altered its trees. inotify will report such accesses to us after all, hence we could hook this up nicely. 

I mean, given that docker, lxc, all currently muck around in the cgroup root even though they really should not and they to some level get away with it, strictly enforcing access here is problematic, we'd break everybody's setups in one step...

binding this to a single PID wouldn't really work i figure btw. That's because in systemd the child we fork off that'll become the main service process will add itself to the right cgroup before exece()ing, in order to guarantee correct membership before the execve() is executed. And that means the child's PID should really be allowed to write to cgroupfs too — up to the point where execve() is executed.

What could work thous if we could bind write cgroup write rights to memberships in specific cgroups. PID 1 and initially all its children run inside the /init.scope/ cgroup after all, and if we could declare that only members of /init.scope/ are permitted to write to the tree up declared delegation points then this would be quite nice.

Reply:
Gating by PID who can write to the limits (memory.usage_in_bytes, etc.) might work though... Even though you might not limit writing to tasks...

Yeah if inotify is an option, maybe just watching and logging from systemd would be a good idea! At least we'd have some traces to point out that the policy was broken...